samplers/parser.go

```diff
@@ -84,6 +85,75 @@ func ValidMetric(sample *UDPMetric) bool {
 	return ret
 }
 
+const expectedFrameVersion uint32 = 0
+const frameHeaderLength = 10
+const maxPacketLength = 16 * 1024 * 1024 // 16MB ought to be enough for anyone (for now)
+
+var (
+	// ErrInvalidFrameLen is an error indicating a malformed frame
+	// on an SSF message stream. All headers are expected to be 10 bytes.
+	ErrInvalidFrameLen = fmt.Errorf("Protocol violation: SSFFrame is not the correct length")
+)
+
+// SSFStreamHeader constructs and returns an SSF frame header for the
+// byte sequence buf.
+func SSFStreamHeader(buf []byte) *ssf.SSFFrameHeader {
+	length := uint32(len(buf))
+	version := expectedFrameVersion
+	return &ssf.SSFFrameHeader{
+		Version: &version,
+		Length:  &length,
+	}
+}
+
+// ReadSSFStream takes a reader and returns an array of unvalidated
+// SSF spans. Callers of ParseSSFStream have to take care to validate
+// the traces contained in each span and to extract any metrics from
+// them.
+func ReadSSFStream(reader io.Reader) ([]*ssf.SSFSpan, error) {
+	frameHdr := ssf.SSFFrameHeader{}
+	// TODO: This probably wants to be a pool
+	header := make([]byte, frameHeaderLength)
+	n, err := reader.Read(header)
+	if err != nil {
+		return []*ssf.SSFSpan{}, err
+	}
+	if n != frameHeaderLength {
+		return []*ssf.SSFSpan{}, ErrInvalidFrameLen
+	}
+	// TODO: This scratchBuff thing feels really dirty. This too wants to be a pool, I'm sure.
+	scratchBuff.Reset()
+	scratchBuff.SetBuf(header)
+	err = scratchBuff.Unmarshal(&frameHdr)
+	if err != nil {
+		return []*ssf.SSFSpan{}, err
+	}
+
+	if *frameHdr.Length > maxPacketLength {
+		return []*ssf.SSFSpan{}, fmt.Errorf("Packet too long: %d > %d (max)", *frameHdr.Length, maxPacketLength)
+	}
+	if *frameHdr.Version != expectedFrameVersion {
+		return []*ssf.SSFSpan{},
+			fmt.Errorf("Protocol violation: SSFFrame has unrecognized version %d, expected %d", *frameHdr.Version, expectedFrameVersion)
+	}
+	packet := make([]byte, *frameHdr.Length)
+	n, err = reader.Read(packet)
+	if err != nil {
+		return []*ssf.SSFSpan{}, err
+	}
+	if uint32(n) != *frameHdr.Length {
+		return []*ssf.SSFSpan{}, fmt.Errorf("Protocol violation: Frame header says packet is %d bytes long, only read %d", frameHdr.Length, n)
+	}
+	scratchBuff.Reset()
+	scratchBuff.SetBuf(packet)
+	spanColl := ssf.SSFSpanCollection{}
+	err = scratchBuff.Unmarshal(&spanColl)
+	if err != nil {
+		return []*ssf.SSFSpan{}, err
+	}
+	return spanColl.Spans, nil
```

Comment:
I wasn't originally thinking we'd put the collection in here, since we could just length prefix each span. Buuuut, it's certainly way less work to do it this way though, eh? :)

Reply:
Yeah, I'm of two minds about this: 

On one hand, if we make clients wrap single spans in Collections unnecessarily, that's a lot of additional weight (also, the reading gets more complicated and full of unnecessary loops on the fast path).

On the other, this is just about what "I want to send loads of spans" should look like.